File: src/KPTech.WS.Mvu.FSharp.Extensions/ExtensionTypes.fs
// ---------------------------------------------------------------------------
// Copyright (c) 2019, Zoltan Podlovics, KP-Tech Kft. All Rights Reserved.
//
// Licensed under the Apache License, Version 2.0. See LICENSE.md in the 
// project root for license information.
// ---------------------------------------------------------------------------

namespace KPTech.WS.Mvu.FSharp.Extensions

open System.Runtime.CompilerServices

open WebSharper
open WebSharper.UI
open WebSharper.Mvu

[<Extension; Sealed; JavaScript>]
type RenderExtension =
            
    [<Extension; Inline>]
    static member FromRenderFunc(func: System.Func<System.Action<'Message>,View<'Model>,'Rendered>) : Dispatch<'Message> -> View<'Model> -> 'Rendered = 
        let f (d: Dispatch<'Message>) (v: View<'Model>) : 'Rendered =
            let dispatchAction : System.Action<'Message> = System.Action<'Message>(fun msg -> d msg)
            func.Invoke(dispatchAction, v)
        f

File: src/KPTech.WS.Mvu.CSharp.Extensions/ExtensionTypes.cs
// Copyright (c) 2019, Zoltan Podlovics, KP-Tech Kft. All Rights Reserved.
//
// Licensed under the Apache License, Version 2.0. See LICENSE.md in the 
// project root for license information.
// ---------------------------------------------------------------------------

/*
using System;
using System.Diagnostics;
using System.Runtime.CompilerServices;
using Microsoft.FSharp.Core;
using WebSharper;
using WebSharper.UI;

namespace KPTech.WS.Mvu.CSharp.Extensions
{
    [Serializable]
    [Sealed]
    [JavaScript]
    internal static class ExtensionTypes
    {
        [Serializable]
        private sealed class DispatchF<TMessage>
        {
            private readonly FSharpFunc<TMessage, Unit> _f;

            public DispatchF (FSharpFunc<TMessage, Unit> f)
            {
                _f = f;
            }

            internal void Invoke (TMessage msg)
            {
                _f.Invoke (msg);
            }
        }

        [Serializable]
        internal sealed class RenderF<TMessage, TModel, TRendered>: Microsoft.FSharp.Core.OptimizedClosures.FSharpFunc<FSharpFunc<TMessage,Unit>,View<TModel>,TRendered>
        {
            private readonly Func<Action<TMessage>, View<TModel>, TRendered> _func;

            [CompilerGenerated]
            [DebuggerNonUserCode]
            internal RenderF (Func<Action<TMessage>, View<TModel>, TRendered> func)
            {
                _func = func;
            }

            public override TRendered Invoke (FSharpFunc<TMessage, Unit> d, View<TModel> v)
            {
                Action<TMessage> action = new DispatchF<TMessage> (d).Invoke;
                return _func.Invoke (action, v);
            }
        }
}
}

*/